tree.go

```diff
@@ -1254,22 +1168,36 @@ func (n *LeafNode) GetProofItems(keys keylist) (*ProofElements, []byte, [][]byte
 	return pe, esses, poass
 }
 
+// Serialize serializes a LeafNode.
+// The format is: <nodeType><stem><bitlist><c1comm><c2comm><children...>
 func (n *LeafNode) Serialize() ([]byte, error) {
-	var bitlist [32]byte
-	children := make([]byte, 0, NodeWidth*32)
+	// Empty value in LeafNode used for padding.
+	var emptyValue [LeafValueSize]byte
+
+	// Create bitlist and store in children LeafValueSize (padded) values.
+	children := make([]byte, 0, NodeWidth*LeafValueSize)
+	var bitlist [bitlistSize]byte
 	for i, v := range n.values {
 		if v != nil {
 			setBit(bitlist[:], i)
 			children = append(children, v...)
-			if len(v) < 32 {
-				padding := make([]byte, 32-len(v))
+			if padding := emptyValue[:32-len(v)]; len(padding) != 0 {
 				children = append(children, padding...)
 			}
 		}
 	}
-	c1bytes := n.c1.Bytes()
-	c2bytes := n.c2.Bytes()
-	return append(append(append(append(append([]byte{leafRLPType}, n.stem[:31]...), bitlist[:]...), c1bytes[:]...), c2bytes[:]...), children...), nil
+
+	// Create the serialization.
+	baseSize := nodeTypeSize + StemSize + bitlistSize + 2*SerializedPointSize
+	result := make([]byte, baseSize, baseSize+4*32) // Extra pre-allocated capacity for 4 values.
+	result[0] = leafRLPType
+	copy(result[leafSteamOffset:], n.stem[:StemSize])
+	copy(result[leafBitlistOffset:], bitlist[:])
+	copy(result[leafC1CommitmentOffset:], n.c1.BytesUncompressed())
+	copy(result[leafC2CommitmentOffset:], n.c2.BytesUncompressed())
+	result = append(result, children...)
```

Comment:
This is pulling another optimization in a [pending PR](https://github.com/gballet/go-verkle/pull/322). It wasn't entirely central to the gains, but it avoids allocations, and that's good. Also, it makes clearer the result construction compared to nested appends.